tool/net/redbean.c

```diff
@@ -4697,6 +4697,162 @@ static int LuaGetHttpReason(lua_State *L) {
+    return luaL_argerror(L, 1, "can't serialize value of this type");
+  }
+  return 0;
+}
+
+static void EscapeLuaString(char *s, size_t len, char **buf) {
+  static char OCT[8] = "01234567";
+  char escape[4] = "\\000";
+  appendd(buf, "\"", 1);
+  for (size_t i = 0; i < len; i++) {
+    if (s[i] == '\\' || s[i] == '\"' || s[i] == '\n' ||
+        s[i] == '\0' || s[i] == '\r') {
+      escape[1] = OCT[(s[i] >> 6) & 7];
+      escape[2] = OCT[(s[i] >> 3) & 7];
+      escape[3] = OCT[(s[i] >> 0) & 7];
+      appendd(buf, escape, 4);
+    } else {
+      appendd(buf, s+i, 1);
+    }
+  }
+  appendd(buf, "\"", 1);
+}
+
+static int EncodeLuaData(lua_State *L, char **buf, int level, char *numformat) {
+  size_t idx = -1;
+  size_t tbllen, buflen, slen;
+  char *s;
+  int ktype;
+  int t = lua_type(L, idx);
+  if (level < 0) return luaL_argerror(L, 1, "too many nested tables");
+  if (LUA_TSTRING == t) {
+    s = lua_tolstring(L, idx, &slen);
+    EscapeLuaString(s, slen, buf);
+  } else if (LUA_TNUMBER == t) {
+    // TODO: what if int64_t isn't representable as double
+    appendf(buf, numformat, lua_tonumber(L, idx));
+  } else if (LUA_TBOOLEAN == t) {
+    appends(buf, lua_toboolean(L, idx) ? "true" : "false");
+  } else if (LUA_TTABLE == t) {
+    appendd(buf, "{", 1);
+    int i = 0;
+    lua_pushnil(L); // push the first key
+    while (lua_next(L, -2) != 0) {
+      ktype = lua_type(L, -2);
+      if (ktype == LUA_TTABLE)
+        return luaL_argerror(L, 1, "can't serialize key of this type");
+      if (i++ > 0) appendd(buf, ",", 1);
+      if (ktype != LUA_TNUMBER || floor(lua_tonumber(L, -2)) != i) {
+        appendd(buf, "[", 1);
+        lua_pushvalue(L, -2); // table/-4, key/-3, value/-2, key/-1
+        EncodeLuaData(L, buf, level, numformat);
+        lua_remove(L, -1); // remove copied key: table/-3, key/-2, value/-1
+        appendd(buf, "]=", 2);
+      }
+      EncodeLuaData(L, buf, level-1, numformat);
+      lua_pop(L, 1); // table/-2, key/-1
+    }
+    // stack: table/-1, as the key was popped by lua_next
+    appendd(buf, "}", 1);
+  } else if (LUA_TNIL == t)
+    appendd(buf, "nil", 3);
+  else {
+    return luaL_argerror(L, 1, "can't serialize value of this type");
+  }
+  return 0;
+}
+
+static int LuaEncodeSmth(lua_State *L, int Encoder(lua_State *, char **, int, char *)) {
+  int useoutput = false;
+  int maxdepth = 64;
+  char *numformat = "%.14g";
+  char *p = 0;
+
+  if (lua_istable(L, 2)) {
+    lua_settop(L, 2);  // discard any extra arguments
+    lua_getfield(L, 2, "useoutput");
+    // ignore useoutput outside of request handling
+    if (ishandlingrequest && lua_isboolean(L, -1)) useoutput = lua_toboolean(L, -1);
+    lua_getfield(L, 2, "maxdepth");
+    maxdepth = luaL_optinteger(L, -1, maxdepth);
+    lua_getfield(L, 2, "numformat");
+    numformat = luaL_optstring(L, -1, numformat);
+  }
+  lua_settop(L, 1); // keep the passed argument on top
+  (void)Encoder(L, useoutput ? &outbuf : &p, maxdepth, numformat);
```

Comment:
Isn't the `(void)` statement thing only really necessary if we're compiling with stricter compiler warnings? Within the monorepo .c files we officially speaking only need to cater to one vendored compiler with one defined configuration. Code with an ultra-source-portable audience is probably better regularly exported to a separate repo, e.g. bestline.

Reply:
Removed.